plutus-core/plutus-ir/src/PlutusIR/Transform/LetFloatIn.hs

```diff
@@ -0,0 +1,329 @@
+{-# LANGUAGE BangPatterns      #-}
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TupleSections     #-}
+{-# LANGUAGE ViewPatterns      #-}
+
+{- | Float bindings inwards.
+
+This usually reduces the evaluation cost of the program, but can sometimes lead to
+increased cost. However, the latter is uncommon and the additional cost is almost
+always small compared to the total cost.
+-}
+module PlutusIR.Transform.LetFloatIn (floatTerm) where
+
+import PlutusCore qualified as PLC
+import PlutusCore.Builtin qualified as PLC
+import PlutusCore.Name qualified as PLC
+import PlutusIR
+import PlutusIR.Purity (isPure)
+import PlutusIR.Transform.Rename ()
+
+import Control.Lens hiding (Strict)
+import Control.Monad ((<=<))
+import Data.Set (Set)
+import Data.Set qualified as Set
+
+{- Note [Which bindings can be floated inwards?]
+
+Moving a lazy binding obviously doesn't change the semantics of the program, since the RHS of
+a lazy binding is always evaluated whenever needed, and only evaluated when needed, regardless
+of where the binding is placed. Lazy bindings therefore can be floated inwards.
+
+Why, one might ask, does floating lazy bindings inwards save cost, if the RHS is always
+evaluated the same number of times? Because when we move a lazy binding `let x (nonstrict) = rhs`,
+what we are really moving around is `delay rhs` and lambda applications, and neither
+`delay rhs` nor lambda application is free - they incur additional `Delay`, `LamAbs`
+and `Apply` steps in the CEK machine. Thus cost can be saved by floating bindings into
+unevaluated branches.
```

Comment:
Coming back to this... we don't move strict non-pure let bindings, so that suggests that we _only_ move bindings such that the difference comes from the overhead cost. So... _all_ the cost savings must be coming from these "trivial" reductions! Which is a) really surprising and b) makes the argument that we don't need to worry about repeating them more questionable.

Or have I got this wrong :thinking:

Reply:
let-bindings are much more expensive than they need to be AFAICT